merlin-driver/src/main/java/gov/nasa/jpl/aerie/merlin/driver/DirectiveTypeRegistry.java

```diff
@@ -0,0 +1,61 @@
+package gov.nasa.jpl.aerie.merlin.driver;
+
+import gov.nasa.jpl.aerie.merlin.protocol.driver.DirectiveTypeRegistrar;
+import gov.nasa.jpl.aerie.merlin.protocol.model.MissionModelFactory;
+import gov.nasa.jpl.aerie.merlin.protocol.model.TaskSpecType;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Function;
+
+public record DirectiveTypeRegistry<Registry, Model>(
+    Map<String, TaskSpecType<Model, ?, ?>> taskSpecTypes,
+    Registry registry)
+{
+  public static <Registry, Model>
+  DirectiveTypeRegistry<Registry, Model> extract(final MissionModelFactory<Registry, Model> factory) {
+    return extract(factory::buildRegistry);
+  }
+
+  public static <Registry, Model>
+  DirectiveTypeRegistry<Registry, Model> extract(final Function<DirectiveTypeRegistrar<Model>, Registry> scope) {
+    final var builder = new Builder<Model>();
+    final var registry = scope.apply(builder);
+    return new DirectiveTypeRegistry<>(builder.taskSpecTypes, registry);
+  }
+
+  public <Input, Output>
+  TaskSpecType<Model, Input, Output> lookup(final DirectiveTypeId<Input, Output> id) {
+    final TaskSpecType<Model, ?, ?> rawDirectiveType = this.taskSpecTypes.get(id.name);
+
+    @SuppressWarnings("unchecked")
+    final var directiveType = (TaskSpecType<Model, Input, Output>) rawDirectiveType;
+
+    return directiveType;
+  }
+
+  private static final class Builder<Model> implements DirectiveTypeRegistrar<Model> {
+    private final Map<String, TaskSpecType<Model, ?, ?>> taskSpecTypes = new HashMap<>();
+
+    @Override
+    public <Input, Output> DirectiveTypeId<Input, Output> registerDirectiveType(
+        final String name,
+        final TaskSpecType<Model, Input, Output> taskSpecType)
+    {
+      // TODO: Throw an exception if this name already has already been registered.
+      this.taskSpecTypes.put(name, taskSpecType);
+      return new DirectiveTypeId<>(name);
+    }
+  }
+
+  public static final class DirectiveTypeId<Input, Output>
+      implements gov.nasa.jpl.aerie.merlin.protocol.driver.DirectiveTypeId<Input, Output>
+  {
+    public final String name;
+
+    private DirectiveTypeId(final String name) {
+      this.name = Objects.requireNonNull(name);
+    }
+  }
+}
```

Comment:
My understanding is that we’re using a class instead of a record here because we want to inherit `hashCode` and `equals` from `Object` rather than have them generated by the Java record syntactic sugar. Is that an accurate read on this?

Reply:
Actually, just because we only want the `DirectiveTypeRegistry` to be able to construct these Ids. If you can conjure an Id up out of nowhere, then you have no guarantee that it will correspond to a directive type in a registry -- nor that the Input and Output types you construct it with will line up with a directive type if one by the given name *does* exist.

Reply:
It's not a foolproof prevention mechanism -- you could just create a registry, register stuff to it to get some IDs, then *query those IDs at a different registry*, but it sets a significantly higher bar.

Reply:
Maybe it would be best to make these IDs non-static, so you can query the directive through the ID itself. That might be safer. Hmmmm.